sw/device/tests/spi_device_smoketest.c

```diff
@@ -0,0 +1,402 @@
+                    expected_irqs[kDifSpiDeviceIrqGenericRxWatermark],
+                kTimeoutUs);
+  LOG_INFO("spi_device rx_watermark interrupt fired.");
+
+  // When 128 bytes received in RX_FIFO from SPI_HOST,
+  // read out and compare against the expected data.
+  IBEX_SPIN_FOR(fired_irqs[kDifSpiDeviceIrqGenericRxWatermark], kTimeoutUs);
+  size_t bytes_recved = 0;
+  uint8_t spi_device_rx_data[kSpiDeviceDatasetSize];
+  CHECK_DIF_OK(dif_spi_device_recv(spi_device, spi_device_rx_data,
+                                   kSpiDeviceDatasetSize, &bytes_recved));
+  CHECK(bytes_recved == kSpiDeviceDatasetSize,
+        "spi_device rx_fifo recvd bytes mismatched: {act: %d, exp: %d}",
+        bytes_recved, kSpiDeviceDatasetSize);
+  LOG_INFO("received %d bytes from spi_device rx_fifo.", bytes_recved);
+  // Check data consistency.
+  LOG_INFO("Checking the received spi_host rx_fifo data for consistency.");
+  CHECK_ARRAYS_EQ(spi_device_rx_data, exp_spi_device_rx_data,
+                  kSpiDeviceDatasetSize);
+}
+
+static void tx_underflow_test(dif_spi_device_handle_t *spi_device) {
+  LOG_INFO("%s", __func__);
+
+  CHECK_DIF_OK(dif_spi_device_set_irq_levels(
+      spi_device, /*rx=*/kSpiDeviceDatasetSize * 2, /*tx=*/1));
+  expected_irqs[kDifSpiDeviceIrqGenericTxUnderflow] = true;
+  expected_irqs[kDifSpiDeviceIrqGenericRxWatermark] = false;
+  // Note: The message bellow is used for syncronization with the host.
+  LOG_INFO("SYNC: TX_FIFO UNDERFLOW READY");
+  IBEX_SPIN_FOR(fired_irqs[kDifSpiDeviceIrqGenericTxUnderflow], kTimeoutUs);
+  LOG_INFO("spi_device tx underflow fired.");
+}
+
+static void rx_full_test(dif_spi_device_handle_t *spi_device) {
+  LOG_INFO("%s", __func__);
+
+  flush_fifo(spi_device);
+  CHECK_DIF_OK(dif_spi_device_set_irq_levels(spi_device, /*rx=*/1024,
+                                             /*tx=*/kSpiDeviceDatasetSize));
+  expected_irqs[kDifSpiDeviceIrqGenericRxWatermark] = false;
+  expected_irqs[kDifSpiDeviceIrqGenericRxFull] = true;
+  // Note: The message bellow is used for syncronization with the host.
+  LOG_INFO("SYNC: RX_FIFO FULL READY");
+  // After RX SRAM FIFO full, expect RX async FIFO overflow irq.
+  IBEX_SPIN_FOR(fired_irqs[kDifSpiDeviceIrqGenericRxFull] &&
+                    !fired_irqs[kDifSpiDeviceIrqGenericRxOverflow],
+                kTimeoutUs);
+  LOG_INFO("Spi_device rx_fifo full interrupt fired.");
+}
+
+static void rx_overflow_test(dif_spi_device_handle_t *spi_device) {
+  LOG_INFO("%s", __func__);
+
+  expected_irqs[kDifSpiDeviceIrqGenericRxFull] = false;
+  expected_irqs[kDifSpiDeviceIrqGenericRxOverflow] = true;
+  // Note: The message bellow is used for syncronization with the host.
+  LOG_INFO("SYNC: RX_FIFO OVERFLOW READY");
+  // After RX SRAM FIFO full, expect RX async FIFO overflow irq.
+  IBEX_SPIN_FOR(fired_irqs[kDifSpiDeviceIrqGenericRxOverflow], kTimeoutUs);
+  LOG_INFO("Spi_device rx_fifo overflow interrupt fired.");
```

Comment:
Should the function drain the FIFO after the test just to be in a good known state?

Reply:
Doesn't hurt, done.